peps/pep-0747.rst

```diff
@@ -0,0 +1,903 @@
+
+      -  Examples: pydantic.\ `TypeAdapter(T).validate_python`_,
+         mashumaro.\ `JSONDecoder(T).decode`_
+
+.. _structure: https://github.com/python-attrs/cattrs/blob/5f5c11627a7f67a23d6212bc7df9f96243c62dc5/src/cattrs/converters.py#L332-L334
+.. _checkcast: https://github.com/davidfstr/trycast#checkcast-api
+.. _load: https://ltworf.github.io/typedload/
+.. _TypeAdapter(T).validate_python: https://stackoverflow.com/a/61021183/604063
+.. _JSONDecoder(T).decode: https://github.com/Fatal1ty/mashumaro?tab=readme-ov-file#usage-example
+
+-  Typed field definitions:
+
+   -  Pattern:
+
+      ::
+
+        class Field[T]:
+            value_type: TypeExpr[T]
+
+   -  Examples: attrs.\ `make_class`_, dataclasses.\ `make_dataclass`_, `openapify`_
+
+.. _make_class: https://www.attrs.org/en/stable/api.html#attrs.make_class
+.. _make_dataclass: https://github.com/python/typeshed/issues/11653
+.. _openapify: https://github.com/Fatal1ty/openapify/blob/c8d968c7c9c8fd7d4888bd2ddbe18ffd1469f3ca/openapify/core/models.py#L16
+
+The survey also identified some introspection functions that take
+annotation expressions as input using plain ``object``\ s which would
+*not* gain functionality by marking those inputs as ``TypeExpr[]``:
+
+-  General introspection operations:
+
+   -  Pattern: ``def get_annotation_info(maybe_annx: object) -> object``
+   -  Examples: typing.{`get_origin`_, `get_args`_},
+      `typing_inspect`_.{is_*_type, get_origin, get_parameters}
+
+.. _get_origin: https://docs.python.org/3/library/typing.html#typing.get_origin
+.. _get_args: https://docs.python.org/3/library/typing.html#typing.get_args
+.. _typing_inspect: https://github.com/ilevkivskyi/typing_inspect?tab=readme-ov-file#readme
+
+
+Specification
+=============
+
+A ``TypeExpr`` represents a `type expression`_ such as ``Optional[str]``,
+``Union[int, str]``, or ``MyTypedDict``. A ``TypeExpr`` type is written as
+``TypeExpr[T]`` where ``T`` is a type or a type variable. It can also be
+written without brackets as just ``TypeExpr``, in which case a type
+checker should apply its usual type inference mechanisms to determine
+the type of its argument, possibly ``Any``.
+
+Type expressions include more values than `annotation expressions`_ but
+fewer values than *class objects* (``type[C]``):
+
+::
+
+   +----------------------------------+ 
+   | +------------------------------+ | 
+   | | +-------------------------+  | | 
+   | | | +---------------------+ |  | | 
+   | | | | Class object        | |  | | = type[C]        
+   | | | +---------------------+ |  | | 
+   | | | Type expression object  |  | | = TypeExpr[T]  <-- new!
+   | | +-------------------------+  | | 
+   | | Annotation expression object | | 
+   | +------------------------------+ | 
+   | Object                           | = object         
+   +----------------------------------+ 
+
+
+Using TypeExprs
+---------------
+
+A ``TypeExpr`` type may be used in any context where a type annotation is
```

Comment:
This can be written more concisely by saying that ``TypeExpr`` itself is a new kind of type expression, which implies that it is valid anywhere a type expression is expected. The term "type annotation" in this line should at least be rewritten to "type expression". `TypeExpr` should also be valid in other type expression contexts that are not annotations, such as in `cast()`.